### ISSUE TITLE
Cactus integration test for releases.

### ISSUE THREAD

**Issue by ejacox** [DataBiosphere/toil#1739]
Clean up and shorten existing release test based on toil-rna-seq and add a test for cactus. These could go into an examples directory.

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-175)
┆Issue Number: TOIL-175

**Comment by adamnovak** [comment]
I'm coopting this issue a bit to represent a Cactus integration test. We want one that we can run and time on Kubernetes to measure the effectiveness of turning on caching in that environment.

**Comment by adamnovak** [comment]
So far, I've tried and failed to run Cactus single-machine on Kubernetes, manually, before putting together a Gitlab test to do it against the Kubernetes batch system. I think we need the `--binariesMode singularity` flag when running with the Kubernetes batch system, because the Toil appliance doesn't run its own nested Docker, so I'm passing that here too. But Cactus doesn't seem to know it needs to use user-mode namespaces to make Singularity work:

```
INFO:cactus.shared.common:Running the command ['singularity', '--silent', 'run', '/root/cactus/tree/cactus.img', 'cactus_blast_chunkSequences', 'INFO', '3000000', '0', 'tirc1j58t/tmp_StfyAIoWjF', 'tmp4chpzljm.tmp']
ERROR  : Failed to create mount namespace: Operation not permitted
```

Maybe there's another flag I am missing?

**Comment by adamnovak** [comment]
It looks like Cactus may onlu use user-mode namespaces when not using a file job store. See here:

\https://github.com/ComparativeGenomicsToolkit/cactus/blob/6577968dbc13632cee289f5cd067ce626f1f1f87/src/cactus/shared/common.py#L1007

And here:

https://github.com/ComparativeGenomicsToolkit/cactus/blob/6577968dbc13632cee289f5cd067ce626f1f1f87/src/cactus/shared/common.py#L948-L949

**Comment by adamnovak** [comment]
Yep, I need to use the AWS job store. I think with that, and with https://github.com/ComparativeGenomicsToolkit/cactus/pull/239 in to fix Python 3 compatibility for the Singularity user mode code path, this may magically start working.

@glennhickey, when he runs Cactus single machine, does it in the `quay.io/comparative-genomics-toolkit/cactus:v1.0.0` container, so all the binaries are ready to hand and no nested containers are needed. But I don't think that will work when we use the Kubernetes batch system. Maybe we should use it as a runtime reference, though; how fast is the Kubernetes batch system and all its nested containers compared to that?

**Comment by adamnovak** [comment]
OK, I've managed to run the same Cactus evolver mammals mouse-rat test single machine and with the Kubernetes batch system.

Single machine time:

```
real    21m56.016s
user    13m5.935s
sys     2m59.655s
```

Kubernetes time:

```
real    52m11.670s
user    1m26.566s
sys     0m6.534s
```

I did it with a fork of Cactus that logs runtimes, and it looks like all the runtime difference is in a few slow jobs. Here's a histogram comparing the time between job issuance and the job coming back, looking only at the high end:

![image](https://user-images.githubusercontent.com/5062495/83085983-5f935c00-a042-11ea-9a0b-5abbad637528.png)

A few jobs get dramatically slower on the Kubernetes batch system. As far as I can tell, they're all `KtServerService` jobs. Here's the log for one:

```
[2020-05-27 23:34:38,690] [MainThread] [I] [toil.worker] ---TOIL WORKER OUTPUT LOG---
[2020-05-27 23:34:38,690] [MainThread] [I] [toil] Running Toil version 4.2.0a1-488bb8a75a23a653710c5c2973e5f310a3719854 on host anovak-toil-d05759f1-7d97-4639-8b54-51dd99c23b29-198-dqhrv.
[2020-05-27 23:34:38,690] [MainThread] [D] [toil] Configuration: {'workflowID': '0508fc97-025e-4d04-a0be-288ac0408fd6', 'workflowAttemptNumber': 0, 'jobStore': 'aws:us-west-2:anovak-toil-cactus-test-kube', 'logLevel': 'DEBUG', 'workDir': None, 'noStdOutErr': False, 'stats': False, 'clean': 'always', 'cleanWorkDir': 'always', 'clusterStats': None, 'restart': False, 'batchSystem': 'kubernetes', 'disableAutoDeployment': False, 'environment': {}, 'statePollingWait': 1, 'maxLocalJobs': 8, 'manualMemArgs': False, 'parasolCommand': 'parasol', 'parasolMaxBatches': 10000, 'scale': 1, 'linkImports': False, 'moveExports': False, 'mesosMasterAddress': '10.244.1.95:5050', 'kubernetesHostPath': None, 'provisioner': None, 'nodeTypes': [], 'nodeOptions': None, 'minNodes': None, 'maxNodes': [10], 'targetTime': 1800, 'betaInertia': 0.1, 'scaleInterval': 60, 'preemptableCompensation': 0.0, 'nodeStorage': 50, 'metrics': False, 'maxPreemptableServiceJobs': 9223372036854775807, 'maxServiceJobs': 9223372036854775807, 'deadlockWait': 3600, 'deadlockCheckInterval': 30, 'defaultMemory': 2147483648, 'defaultCores': 1, 'defaultDisk': 8589934592, 'readGlobalFileMutableByDefault': False, 'defaultPreemptable': False, 'maxCores': 9223372036854775807, 'maxMemory': 9223372036854775807, 'maxDisk': 9223372036854775807, 'retryCount': 5, 'enableUnlimitedPreemptableRetries': False, 'maxJobDuration': 9223372036854775807, 'rescueJobsFrequency': 3600, 'disableCaching': True, 'disableChaining': True, 'disableJobStoreChecksumVerification': False, 'maxLogFileSize': 64000, 'writeLogs': None, 'writeLogsGzip': None, 'writeLogsFromAllJobs': False, 'sseKey': None, 'servicePollingInterval': 60, 'useAsync': True, 'forceDockerAppliance': False, 'runCwlInternalJobsOnWorkers': False, 'statusWait': 3600, 'disableProgress': False, 'debugWorker': False, 'disableWorkerOutputCapture': False, 'badWorker': 0.0, 'badWorkerFailInterval': 0.01, 'cwl': False}
[2020-05-27 23:34:38,690] [MainThread] [D] [toil.deferred] Running for file /var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/deferred/funcop7qs_q_
[2020-05-27 23:34:38,704] [MainThread] [D] [toil.jobStores.aws.jobStore] Loaded job e230b54b-be8b-4065-a8a4-6995f0a83e02
[2020-05-27 23:34:38,704] [MainThread] [D] [toil.worker] Parsed job wrapper
[2020-05-27 23:34:38,704] [MainThread] [D] [toil.worker] Got a command to run: _toil 3625e4f4-8836-4471-8f08-52b8fdaafa6f /root/venv/lib/python3.7/site-packages cactus.pipeline.ktserverToil True
[2020-05-27 23:34:38,704] [MainThread] [D] [toil.job] Loading user module ModuleDescriptor(dirPath='/root/venv/lib/python3.7/site-packages', name='cactus.pipeline.ktserverToil', fromVirtualEnv=True).
[2020-05-27 23:34:45,191] [MainThread] [D] [toil.jobStores.aws.jobStore] Reading FileInfo(fileID='3625e4f4-8836-4471-8f08-52b8fdaafa6f', ownerID='5e1ba609-ea85-4e38-a63d-986e10eba20e', encrypted=False, version='F_ZWUSGJXbvRgvzzE13szdpd0JwhykI1', previousVersion='F_ZWUSGJXbvRgvzzE13szdpd0JwhykI1', content=None, checksum='sha1$cc61cebc2e1...f8de7da4afa6b2ab3', _numContentChunks=0) bound to <toil.jobStores.aws.jobStore.AWSJobStore object at 0x7f16c29c4050> into '/tmp/tmpj2ps7rri/job'.
[2020-05-27 23:34:45,900] [MainThread] [D] [toil.jobStores.aws.jobStore] Starting sha1 checksum to match cc61cebc2e11f41a601ff0ff8de7da4afa6b2ab3
[2020-05-27 23:34:45,900] [MainThread] [D] [toil.jobStores.aws.jobStore] Updating checksum with 3935 bytes
[2020-05-27 23:34:45,900] [MainThread] [D] [toil.jobStores.aws.jobStore] Completed checksum with hash cc61cebc2e11f41a601ff0ff8de7da4afa6b2ab3 vs. expected cc61cebc2e11f41a601ff0ff8de7da4afa6b2ab3
[2020-05-27 23:34:45,903] [MainThread] [D] [toil.deferred] Running orphaned deferred functions
[2020-05-27 23:34:45,903] [MainThread] [D] [toil.deferred] Running job
[2020-05-27 23:34:45,904] [MainThread] [D] [toil.job] Loading service module ModuleDescriptor(dirPath='/root/venv/lib/python3.7/site-packages', name='cactus.pipeline.ktserverToil', fromVirtualEnv=True).
[2020-05-27 23:34:48,170] [Thread-5  ] [D] [toil.jobStores.aws.jobStore] Inlining content of 0 bytes
[2020-05-27 23:34:48,250] [MainThread] [D] [toil.jobStores.aws.jobStore] Created FileInfo(fileID='fc62a9e0-aad8-4789-8b8d-87d19c8c45d0', ownerID=None, encrypted=False, version='', previousVersion='', content=b'', checksum='', _numContentChunks=1) bound to <toil.jobStores.aws.jobStore.AWSJobStore object at 0x7f16c29c4050>.
[2020-05-27 23:34:48,265] [MainThread] [D] [toil.jobStores.aws.jobStore] Reading FileInfo(fileID='fc62a9e0-aad8-4789-8b8d-87d19c8c45d0', ownerID='None', encrypted=False, version='', previousVersion='', content=b'', checksum='', _numContentChunks=1) bound to <toil.jobStores.aws.jobStore.AWSJobStore object at 0x7f16c29c4050> into '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6/tmpnpyyj511.tmp'.
[2020-05-27 23:34:48,265] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:34:48,265] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:34:48,265] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'netstat', '-tuplen']
WARNING: skipping mount of /etc/localtime: no such file or directory
[2020-05-27 23:34:50,271] [MainThread] [D] [toil.lib.bioio] Detected ports in use: set()
[2020-05-27 23:34:50,375] [MainThread] [I] [cactus.shared.common] Work dirs: {'/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6/twn680bf9', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6'}
[2020-05-27 23:34:50,375] [MainThread] [I] [cactus.shared.common] Docker work dir: /var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6
[2020-05-27 23:34:50,376] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktserver', '-port', '20079', '-ls', '-tout', '200000', '-th', '64', '-bgs', 'twn680bf9/snapshot', '-bgsc', 'lzo', '-bgsi', '1000000', '-log', 'tmps9219gn0.tmp', ':#opts=ls#bnum=30m#msiz=50g#ktopts=p']
WARNING: skipping mount of /etc/localtime: no such file or directory
[2020-05-27 23:34:52,470] [MainThread] [I] [toil.lib.bioio] Ktserver running.
[2020-05-27 23:34:52,471] [MainThread] [I] [toil.lib.bioio] Ktserver running.
[2020-05-27 23:34:52,588] [Thread-6  ] [D] [toil.jobStores.aws.jobStore] Inlining content of 159 bytes
[2020-05-27 23:34:52,670] [MainThread] [I] [toil.lib.bioio] Ktserver running.
[2020-05-27 23:34:52,670] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:34:52,670] [MainThread] [D] [toil.jobStores.aws.jobStore] Wrote FileInfo(fileID='28a0a831-f3b2-4275-b2fc-00b1eb78e145', ownerID='None', encrypted=False, version='', previousVersion='', content=b'\x80\x04\x9...onf>\n\t\x94.', checksum='', _numContentChunks=1) bound to <toil.jobStores.aws.jobStore.AWSJobStore object at 0x7f16c29c4050> from stream.
[2020-05-27 23:34:52,670] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:34:52,670] [MainThread] [D] [toil.job] Removing the start jobStoreID to indicate that establishment of the service
[2020-05-27 23:34:52,670] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:36:01,530] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:36:01,530] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:36:01,531] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:37:20,970] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:37:20,970] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:37:20,970] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:38:27,344] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:38:27,344] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:38:27,344] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:39:34,430] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:39:34,430] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:39:34,430] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:40:44,370] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:40:44,370] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:40:44,371] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:41:49,269] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:41:49,270] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:41:49,270] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:42:54,270] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:42:54,270] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:42:54,270] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:44:00,831] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:44:00,831] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:44:00,831] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:45:04,769] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:45:04,770] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:45:04,770] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:46:10,170] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:46:10,170] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:46:10,170] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:47:29,630] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:47:29,630] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:47:29,631] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:48:37,530] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:48:37,530] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:48:37,531] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:49:42,603] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:49:42,603] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:49:42,603] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:50:48,831] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:50:48,831] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:50:48,831] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:51:55,079] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:51:55,079] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:51:55,079] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:53:02,230] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:53:02,230] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:53:02,269] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
[2020-05-27 23:53:59,191] [MainThread] [D] [toil.job] Detected that the terminate jobStoreID has been removed so exiting
[2020-05-27 23:53:59,207] [MainThread] [D] [toil.job] Service is done
[2020-05-27 23:53:59,207] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:53:59,207] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:53:59,207] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'set', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE', '1']
WARNING: skipping mount of /etc/localtime: no such file or directory
[2020-05-27 23:54:10,130] [MainThread] [I] [cactus.shared.common] Work dirs: set()
[2020-05-27 23:54:10,130] [MainThread] [I] [cactus.shared.common] Docker work dir: .
[2020-05-27 23:54:10,131] [MainThread] [I] [cactus.shared.common] Running the command ['singularity', '-q', 'exec', '-u', '-B', '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6:/mnt', '--pwd', '/mnt', '/var/lib/toil/singularity-cache/toil/0d6287f429ecb80cbdcd8de224eca432a363f4852ea4d66d2047979bb17ccfa9.sandbox', 'ktremotemgr', 'get', '-port', '20079', '-host', '10.244.1.53', 'TERMINATE']
1
[2020-05-27 23:54:19,998] [MainThread] [D] [toil.jobStores.aws.jobStore] Starting sha1 checksum to match None
[2020-05-27 23:54:19,998] [MainThread] [D] [toil.jobStores.aws.jobStore] Updating checksum with 488596 bytes
[2020-05-27 23:54:19,999] [MainThread] [D] [toil.jobStores.aws.jobStore] Completed checksum with hash f6b7ac3b6b643e649a60d43eb3356cc34712b6d9 vs. expected None
[2020-05-27 23:54:20,288] [MainThread] [D] [toil.jobStores.aws.jobStore] Wrote FileInfo(fileID='fc62a9e0-aad8-4789-8b8d-87d19c8c45d0', ownerID='None', encrypted=False, version='J_U3UR9V4zq5QD3pFcl2LvXQR3wv5JEn', previousVersion='J_U3UR9V4zq5QD3pFcl2LvXQR3wv5JEn', content=None, checksum='sha1$f6b7ac3b6b6...eb3356cc34712b6d9', _numContentChunks=0) bound to <toil.jobStores.aws.jobStore.AWSJobStore object at 0x7f16c29c4050> from path '/var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql/a4c223dd-d4b6-4941-a580-b985eb8043c6/twn680bf9/snapshot/00000000.ktss'.
[2020-05-27 23:54:27,785] [MainThread] [D] [toil.fileStores.abstractFileStore] LOG-TO-MASTER: Job 3625e4f4-8836-4471-8f08-52b8fdaafa6f used 0.01% (488.0 KB [499712B] used, 8.0 GB [8589934592B] requested) at the end of its run.
[2020-05-27 23:54:27,786] [MainThread] [D] [toil.deferred] Running own deferred functions
[2020-05-27 23:54:27,786] [MainThread] [D] [toil.deferred] Out of deferred functions!
[2020-05-27 23:54:27,786] [MainThread] [D] [toil.deferred] Running orphaned deferred functions
[2020-05-27 23:54:27,786] [MainThread] [D] [toil.jobStores.aws.jobStore] Updating job e230b54b-be8b-4065-a8a4-6995f0a83e02
[2020-05-27 23:54:27,871] [MainThread] [D] [toil.worker] Stopping running chain of jobs: length of stack: 0, services: 0, checkpoint: False
[2020-05-27 23:54:27,871] [MainThread] [I] [toil.worker] Worker log can be found at /var/lib/toil/node-0508fc97-025e-4d04-a0be-288ac0408fd6-2d0b4690352c4d58bf00d65e964f0248/tmp0o2x3iql. Set --cleanWorkDir to retain this log
[2020-05-27 23:54:27,871] [MainThread] [I] [toil.worker] Finished running the chain of jobs on this node, we ran for a total of 1189.167089 seconds
```

I can't tell if this is taking a long time because something is wrong with it, or because the leader is taking its time before getting around to issuing jobs that need to be issued. The `KtServerService` jobs are going to take as long to run as all the jobs that depend on them, and it might just be waiting around for those other jobs.

I'm attaching the full logs for comparison.

[log.single.txt](https://github.com/DataBiosphere/toil/files/4692620/log.single.txt)

[log.kube.txt](https://github.com/DataBiosphere/toil/files/4692618/log.kube.txt)

@glennhickey Is that what a `KtServerService` log should look like? It doesn't look much different form the single machine ones to me.

**Comment by diekhans** [comment]
Ugg

for a data point, you might ask Mobin to do the same experiment with his redis fork.

**Comment by adamnovak** [comment]
After thinking about this, I think that the KtServerService must be a red herring. Cactus is clearly running jobs against it the whole time it is up; it must be changes in the runtime or latency (or scheduling) of the non-server jobs that really drive the slowdown.

**Comment by adamnovak** [comment]
OK, I've dug into the Cactus jobs, and here are the jobs that take the most additional total time on Kubernetes vs. single machine, which aren't `KtServerService`:

```
CactusSetReferenceCoordinatesDownWrapper	120.773
CactusHalGeneratorRecursion	130.79700000000003
CactusReferenceWrapper	148.18500000000003
CactusHalGeneratorUpWrapper	365.53599999999994
CactusSetReferenceCoordinatesUpWrapper	448.82700000000006
```

These are just the 5 worst offender job types, and they together account for about 20 minutes worth of extra time.

Some of this looks like genuinely weirdly-behaving jobs that we can debug. For example, job 5e9f8f5e-a911-4fa8-8e34-fb645890e1f1 on Kubernetes, an instance of CactusSetReferenceCoordinatesUpWrapper with batch system ID 243, takes 179.576 seconds from issuance to completion on Kubernetes, and spends 175.468331 seconds actually running as tracked by Toil. What I think is the corresponding job on single machine, also with batch system ID 243 (6d87ccd4-0b42-4323-ac7a-1cb233c004a2) only runs for 3.645133 seconds. This job produces an extremely long Cactus log, so I can't see if at the beginning the Kubernetes job spends a long time downloading files or something. It could be that it got stuck downloading a file to the cache.

@glennhickey How do I run Cactus without the very long Cactus binary logs about flowers and so on, so I can see the Toil log messages from the beginning of each job? And do these long/big jobs wait on other things to happen via the services? Or do they just run as fast as they can?

But it looks like a lot of time is going into very short jobs, which on single machine take maybe 2 seconds but on Kubernetes take maybe 6 seconds. Here are all the issued-to-done times for CactusHalGeneratorRecursion jobs on Kubernetes:
```
6.021	CactusHalGeneratorRecursion	1e5a53cd-7552-4c7d-9165-c8af81394bbe
3.777	CactusHalGeneratorRecursion	1e5a53cd-7552-4c7d-9165-c8af81394bbe
7.619	CactusHalGeneratorRecursion	391693db-2721-4178-94e7-f67223ffb087
3.43	CactusHalGeneratorRecursion	391693db-2721-4178-94e7-f67223ffb087
8.677	CactusHalGeneratorRecursion	6c2d8877-df69-4f61-9001-7812674bbc19
4.072	CactusHalGeneratorRecursion	6c2d8877-df69-4f61-9001-7812674bbc19
9.497	CactusHalGeneratorRecursion	6d99b589-a973-4d87-94e4-385b625bd343
3.823	CactusHalGeneratorRecursion	6d99b589-a973-4d87-94e4-385b625bd343
5.679	CactusHalGeneratorRecursion	6f139ecb-c1c8-4b13-b1bb-cd2ab7dcb01a
3.569	CactusHalGeneratorRecursion	6f139ecb-c1c8-4b13-b1bb-cd2ab7dcb01a
7.392	CactusHalGeneratorRecursion	890edf9a-c4de-4f78-ac68-c7efac269fdb
5.461	CactusHalGeneratorRecursion	890edf9a-c4de-4f78-ac68-c7efac269fdb
4.973	CactusHalGeneratorRecursion	8d4187c6-060e-462c-bf02-d75bcaadaaa7
3.424	CactusHalGeneratorRecursion	8d4187c6-060e-462c-bf02-d75bcaadaaa7
6.049	CactusHalGeneratorRecursion	95212219-dad9-4633-8114-e5da0e3baef0
4.798	CactusHalGeneratorRecursion	95212219-dad9-4633-8114-e5da0e3baef0
6.195	CactusHalGeneratorRecursion	968ca691-0816-4c31-b75a-82183eb69304
3.995	CactusHalGeneratorRecursion	968ca691-0816-4c31-b75a-82183eb69304
8.396	CactusHalGeneratorRecursion	9a4aaf8f-5bf9-4849-9849-c76c10a27643
3.636	CactusHalGeneratorRecursion	9a4aaf8f-5bf9-4849-9849-c76c10a27643
5.831	CactusHalGeneratorRecursion	9e185080-8141-4f55-81c0-a51dc3eb89b2
4.403	CactusHalGeneratorRecursion	9e185080-8141-4f55-81c0-a51dc3eb89b2
5.451	CactusHalGeneratorRecursion	a1f31d80-5bef-4143-8916-5d24d23219d9
4.521	CactusHalGeneratorRecursion	a1f31d80-5bef-4143-8916-5d24d23219d9
5.824	CactusHalGeneratorRecursion	aa17e81d-8fef-41e2-9794-a44ab550a0bb
3.112	CactusHalGeneratorRecursion	aa17e81d-8fef-41e2-9794-a44ab550a0bb
7.599	CactusHalGeneratorRecursion	aaf63252-3c30-4801-ba3e-343fcbdeab47
4.458	CactusHalGeneratorRecursion	aaf63252-3c30-4801-ba3e-343fcbdeab47
13.986	CactusHalGeneratorRecursion	bf7d7396-89ba-4b04-a54e-6541d3b6ee64
5.174	CactusHalGeneratorRecursion	bf7d7396-89ba-4b04-a54e-6541d3b6ee64
6.247	CactusHalGeneratorRecursion	c97ce384-c926-4594-937b-a1c0316dfc90
3.374	CactusHalGeneratorRecursion	c97ce384-c926-4594-937b-a1c0316dfc90
6.502	CactusHalGeneratorRecursion	d5d4561c-92a3-4a9f-946f-4ddd7cd6f0e5
3.711	CactusHalGeneratorRecursion	d5d4561c-92a3-4a9f-946f-4ddd7cd6f0e5
```

And single machine:
```
2.283	CactusHalGeneratorRecursion	05ce1bb2-bcdd-4b33-88bf-960f11256743
1.27	CactusHalGeneratorRecursion	05ce1bb2-bcdd-4b33-88bf-960f11256743
2.258	CactusHalGeneratorRecursion	0afc7d59-c5be-46ac-b10e-df6e8dbd3a28
1.328	CactusHalGeneratorRecursion	0afc7d59-c5be-46ac-b10e-df6e8dbd3a28
1.982	CactusHalGeneratorRecursion	1222caa5-5171-438f-ad30-a611ab6f52fe
1.579	CactusHalGeneratorRecursion	1222caa5-5171-438f-ad30-a611ab6f52fe
2.084	CactusHalGeneratorRecursion	1dc94434-cc39-450a-8439-290a40db7747
1.203	CactusHalGeneratorRecursion	1dc94434-cc39-450a-8439-290a40db7747
3.432	CactusHalGeneratorRecursion	3a88abc2-6752-4808-b296-cd5d90b96261
1.242	CactusHalGeneratorRecursion	3a88abc2-6752-4808-b296-cd5d90b96261
2.225	CactusHalGeneratorRecursion	3c7f90a2-e23d-4dac-b4e5-5e0933b39e6c
1.208	CactusHalGeneratorRecursion	3c7f90a2-e23d-4dac-b4e5-5e0933b39e6c
2.16	CactusHalGeneratorRecursion	3ccfa48b-6bf9-448e-9ea3-1131b2481c8b
1.168	CactusHalGeneratorRecursion	3ccfa48b-6bf9-448e-9ea3-1131b2481c8b
2.167	CactusHalGeneratorRecursion	3e7fcf01-8a93-4ead-a5f2-e41acd32691a
1.203	CactusHalGeneratorRecursion	3e7fcf01-8a93-4ead-a5f2-e41acd32691a
2.811	CactusHalGeneratorRecursion	5192fe93-ba0f-4e26-a4c4-228e35c613fc
1.427	CactusHalGeneratorRecursion	5192fe93-ba0f-4e26-a4c4-228e35c613fc
2.17	CactusHalGeneratorRecursion	5e9264fd-0b28-408d-b9bf-10bf3e4c7c25
1.239	CactusHalGeneratorRecursion	5e9264fd-0b28-408d-b9bf-10bf3e4c7c25
2.641	CactusHalGeneratorRecursion	61c8dbdc-d8c0-4244-8a10-014768149999
1.204	CactusHalGeneratorRecursion	61c8dbdc-d8c0-4244-8a10-014768149999
2.497	CactusHalGeneratorRecursion	62817e53-7498-4d53-85c1-35421b84b53b
1.27	CactusHalGeneratorRecursion	62817e53-7498-4d53-85c1-35421b84b53b
2.0	CactusHalGeneratorRecursion	6de2376b-e2f6-47f7-bb93-d7a17328a9b4
1.126	CactusHalGeneratorRecursion	6de2376b-e2f6-47f7-bb93-d7a17328a9b4
2.338	CactusHalGeneratorRecursion	73ff3f80-012e-480f-b6cc-b9a0985c4ad6
1.186	CactusHalGeneratorRecursion	73ff3f80-012e-480f-b6cc-b9a0985c4ad6
1.809	CactusHalGeneratorRecursion	a17cbf24-0c64-4c1d-be12-6d8eb08fd75a
1.077	CactusHalGeneratorRecursion	a17cbf24-0c64-4c1d-be12-6d8eb08fd75a
1.699	CactusHalGeneratorRecursion	eb373f4c-fcc4-4ddd-baa7-e41601c0ab2d
1.162	CactusHalGeneratorRecursion	eb373f4c-fcc4-4ddd-baa7-e41601c0ab2d
2.108	CactusHalGeneratorRecursion	ef0b993d-3f7b-4d7f-a79e-fbf14bd2b714
1.323	CactusHalGeneratorRecursion	ef0b993d-3f7b-4d7f-a79e-fbf14bd2b714
```

We run a lot of these jobs, and if each one takes an extra 5 seconds, it can add up. We run something like 300 jobs, and if each of them takes an extra 5 seconds, that's 25 minutes right there. Looking at Kubernetes job 1e5a53cd-7552-4c7d-9165-c8af81394bbe (6 seconds) vs. what I think is the corresponding single machine job eb373f4c-fcc4-4ddd-baa7-e41601c0ab2d (1.6 seconds), it looks like the extra time goes to:
 * Waiting between job issuance and job start: ~+1 second, makes sense
 * Hot-deploying user module: ~+1 second, makes some sense because user module deployment can't (yet?) use the file store cache.
 * Actual Cactus command execution via Singularity: ~+1 second, doesn't really make sense. Why would the actual Cactus command take longer? Maybe using hostpath-mounted storage is slower than ephemeral storage? Maybe running in a smaller container gets you kicked off the CPU more often to enforce the CPU usage limits?
 * Waiting between job completion on the node and announcement of job end at the leader: ~+1 second, makes sense

I'm not confident that I would be able to chase down and eliminate the extra time here. It may be that the Kubernetes batch system requires each job to be able to tolerate ~4 seconds of extra overhead vs. single machine, without increasing the overall workflow runtime appreciably, in order to be a sensible choice. If we need less overhead than that, we might need to wait for the Toil-internal scheduler.

I will see about turning the evolverMammals Cactus run into a test that just takes a while to run. I'm not sure I can trim the runtime much without digging into the actual important parts of the Cactus workflow.

**Comment by glennhickey** [comment]
cactus gets the loglevel from here: https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/src/cactus/shared/common.py#L77-L82

which gets it from here: https://github.com/DataBiosphere/toil/blob/master/src/toil/lib/bioio.py#L46-L49

and it gets passed to the various `cactus_` commands as the first agurment, for example here:

https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/src/cactus/shared/common.py#L147-L157

I think the bit of extra per-job overhead adding up to explain a lot of the slowness makes sense.  The number of jobs scales at least quadratically with the input size, and the queue gets swamped pretty quickly. 

I do have a hacky work-around that I'm testing that uses double toil:  kubernetes toil to launch a few big jobs, each of which use single machine.  The WDL workflow works in the same way.

**Comment by adamnovak** [comment]
I guess if I want to run Toil with debug log level but the Cactus
binary with less logging, I would need to hack the workflow.

Some of the slowness does seem to be attributable to single jobs.
Maybe that can be debugged away.
